### PR TITLE
Added the ability to set properties in interface builder for FlutterViewController

### PR DESCRIPTION
**Pull request by gaaclarke**
## Description

Added the ability to set properties in interface builder for FlutterViewControllers and their subclasses.  I also added one property that controls if the created engine allows headless execution.

Since our templates use Storyboards we should probably support them better.

## Related Issues

[research into multiple flutters]

## Tests

Covered in other integration tests.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 1]
```diff
@@ -137,7 +141,15 @@ - (instancetype)initWithNibName:(NSString*)nibNameOrNil bundle:(NSBundle*)nibBun
 }
 
 - (instancetype)initWithCoder:(NSCoder*)aDecoder {
-  return [self initWithProject:nil nibName:nil bundle:nil];
+  self = [super initWithCoder:aDecoder];
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:144]: The properties should be pulled off `aDecoder` at initialization (with copies where appropriate)
```obj-c
  _viewOpaque = aDecoder->_viewOpaque;
  _weakFactory = aDecoder->_weakFactory;
  _engine = aDecoder->_engine;
  _flutterView = aDecoder->_flutterView;
  _isDisplayingFlutterUI = [aDecoder decodeObjectForKey:@"isDisplayingFlutterUI"];
  _isHomeIndicatorHidden = [aDecoder decodeObjectForKey:@"isHomeIndicatorHidden"];
  _isPresentingViewControllerAnimating = [aDecoder decodeObjectForKey:@"isPresentingViewControllerAnimating"];
```
not re-set up as if it's a newly initialized object (it's being deserialized).

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 2]
```diff
@@ -111,22 +111,26 @@ - (instancetype)initWithEngine:(FlutterEngine*)engine
...
+}
+
 - (instancetype)initWithProject:(nullable FlutterDartProject*)project
                         nibName:(nullable NSString*)nibName
                          bundle:(nullable NSBundle*)nibBundle {
   self = [super initWithNibName:nibName bundle:nibBundle];
   if (self) {
-    _viewOpaque = YES;
-    _weakFactory = std::make_unique<fml::WeakPtrFactory<FlutterViewController>>(self);
-    _engine.reset([[FlutterEngine alloc] initWithName:@"io.flutter"
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:121]: How is this working?  Where is `_engine` set?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 3]
```diff
@@ -111,22 +111,26 @@ - (instancetype)initWithEngine:(FlutterEngine*)engine
   return self;
 }
 
+- (void)sharedSetupWithProject:(nullable FlutterDartProject*)project {
+  _viewOpaque = YES;
+  _weakFactory = std::make_unique<fml::WeakPtrFactory<FlutterViewController>>(self);
+  _engine.reset([[FlutterEngine alloc] initWithName:@"io.flutter"
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:117]: Isn't `_engine` `nil` here?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 2 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:121]: _engine is now set in `sharedSeutpWithProject`

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 3 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:117]: _engine is a smart pointer, it represents nil until this reset function is called on it.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 1 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:144]: The NSCoder doesn't have a representation of a full object.  It just has what is coming from the xib, so it won't have things like a _weakFactory.  The default initWithCoder will use key-value to set properties that have been specified in the xib, you don't need to create custom readers for them.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 2 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:121]: Oh, to be clear, `awakeFromNib` is calling `sharedSetupWithProject`.  Only after that call are the key-value properties set.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 1 above)
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:144]: I never did grok `UIViewController` or `NSViewController`.  Looked at more docs, you're right.

**Comment by gaaclarke** [comment]
Landing on red, reran the failing test and it passed here: https://ci.chromium.org/raw/build/logs.chromium.org/flutter/led/aaclarke_google.com/6aeea40c8ec6a670a06ab6aed701c523a4ac49cf7c9b04ab4a34367490d483ce/+/annotations?server=chromium-swarm.appspot.com

**Pull request merged**